### PR TITLE
Refactor limit handling

### PR DESCRIPTION
**Pull request by andlaus**
today, we set out to clean up the limit handling code.  this is surprisingly difficult: the problem is that the LIMIT tag does not specify the data type (integer, float, string, bytefield) which the limit is about but the limit obviously needs to know that type in order to do something. The solution taken here is to pass the type of the limit's value to the `from_et()` method as an extra argument. For multiplexer cases where even the type of the limit's value is not known when the limit must be instantiated, (this is only known after the reference to the switch key has been resolved) the method `Limit.set_value_type()` is added.

Andreas Lauser &lt;andreas.lauser@mercedes-benz.com&gt;, on behalf of [MBition GmbH](https://mbition.io/).
[Provider Information](https://github.com/mercedes-benz/foss/blob/master/PROVIDER_INFORMATION.md)

### CODE REVIEW

File: odxtools/compumethods/compuscale.py [hunk 1]
```diff
@@ -64,4 +74,35 @@ def from_et(et_element: ElementTree.Element, doc_frags: List[OdxDocFragment], *,
...
             compu_inverse_value=compu_inverse_value,
             compu_const=compu_const,
-            compu_rational_coeffs=compu_rational_coeffs)
+            compu_rational_coeffs=compu_rational_coeffs,
+            internal_type=internal_type,
+            physical_type=physical_type)
+
+    def applies(self, internal_value: AtomicOdxType) -> bool:
+
+        if self.lower_limit is None and self.upper_limit is None:
```
Comment (kayoub5) [inline odxtools/compumethods/compuscale.py:83]: limits are required?

**Comment by kayoub5** [comment]
Does the limit really need to know its type during creation?

The datatype can be simply passed as argument to `complies_to_upper` and `complies_to_lower`

File: odxtools/compumethods/compuscale.py (hunk 1 above)
Comment (andlaus) [reply odxtools/compumethods/compuscale.py:83]: the XSD says that they are optional for COMPU-SCALE. (this is part of why handling of limits is so awkward...)

**Comment by andlaus** [comment]
> Does the limit really need to know its data type during creation?

not during creation, but as soon as it must check if a value is within the specified limit.

> The datatype can be simply passed as argument to complies_to_upper and complies_to_lower

I considered doing that. It turns out that this implies that the limit value must be converted from string each and every time which a limit is checked (and the code felt quite clunky and "footgunny" because one can easily pass the wrong type to these methods...)

File: odxtools/compumethods/compuscale.py (hunk 1 above)
Comment (kayoub5) [reply odxtools/compumethods/compuscale.py:83]: this is the compu scale class... am I missing something?

Comment (andlaus) [reply odxtools/compumethods/compuscale.py:83]: ah, okay. I suppose that I misread your point: this is supposed to be True?! (did I ever mention that limit handling is surprisingly difficult? anyway: fixed

**Comment by andlaus** [comment]
@kayoub5: Do you think that this can be merged?

**Review by kayoub5 — APPROVED** [review]

**Pull request merged**